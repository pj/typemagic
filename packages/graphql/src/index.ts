export {schema} from './schema';
// export {resolver} from './resolvers';
export {ScalarTypes} from './types';
export {query} from './client';
export {customScalar} from './custom_scalar';
