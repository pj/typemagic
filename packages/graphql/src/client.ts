import { ValidateSchema } from "./schema";
import { GetTypeScalar, IsSchemaScalar, IsTypeScalar, RemovePromise, ScalarTypes, TransformResolverToType } from "./types";
import { ArgumentNode, ASTNode, DocumentNode, FieldNode, GraphQLAbstractType, NameNode, ValueNode, VariableNode } from "graphql";

export type FieldSentinel = {};
export const _: FieldSentinel = {};

class FloatLiteral {
  constructor(public num: number) {}
}

class IntLiteral {
  constructor(public num: number) {}
}

export function floatLiteral(num: number) {
  return new FloatLiteral(num);
}

class Variable<Type> {
  constructor(public name: string) {
  }
}

export function variable<Type = never>(name: string) {
  return new Variable<Type>(name);
}

export type GenerateArgsVariables<Args> =
  {
    args: {
      [Key in keyof Args]: Args[Key] | Variable<Args[Key]>
    }
  }

export type GenerateArgsField<Type, ResolverFunction, Args> =
  [unknown] extends [Args]
    ? Type
    :[undefined] extends [Args]
      ? Type
      : (
          ResolverFunction extends (args: infer Args, root: infer Root, context: infer Context) => any
          ? GenerateArgsVariables<Args>
          : "Should not happen"
        ) & Type

export type GenerateQueryField<Resolver> = 
  [Resolver] extends [
    {
      type: infer Type,
      argsFields?: infer Args,
      resolve?: infer ResolverFunction
    }
  ] 
    ? (
        Type extends {enum: infer Enum, name: infer Name, description?: infer Description}
          ? GenerateArgsField<FieldSentinel, ResolverFunction, Args>
          : IsSchemaScalar<Type> extends true
            ? GenerateArgsField<FieldSentinel, ResolverFunction, Args>
            : Type extends {unionName: infer UnionName, unionTypes: infer UnionTypes}
              ? UnionTypes extends Readonly<unknown[]>
                ?  GenerateArgsField<
                    {
                      __typename: FieldSentinel,
                      on: {
                        [Index in keyof UnionTypes]: "asdf"
                      }
                    }, 
                    ResolverFunction, 
                    Args
                  >
                : "Union types must be an array"
              : Type extends {objectFields: infer ObjectFields}
                ? GenerateArgsField<
                    {
                      [Key in keyof ObjectFields]?:
                        GenerateQueryField<ObjectFields[Key]>
                    }, 
                    ResolverFunction, 
                    Args
                  >
                : "Unable to determine type for client"
      )
    : "Invalid Resolver"

export type GenerateQuery<Schema extends ValidateSchema<Schema, any>> = {
  [Key in keyof Schema['queries']]?: GenerateQueryField<Schema['queries'][Key]>
}

export type GenerateResultField<Schema, Field> =
  [Field] extends [FieldSentinel]
    ? TransformResolverToType<Schema>
    : GenerateResultFields<Schema, Field>

export type GenerateResultFields<Schema, Query> =
{
  [Key in keyof Query]:
    [Key] extends [keyof Schema]
      ? GenerateResultField<Schema[Key], Query[Key]>
      : "Should not happen"
}

export type GenerateResult<Schema extends ValidateSchema<Schema, any>, Query> =
  GenerateResultFields<Schema["queries"], Query>

export type RawReturn<Result> = 
  {
    data: Result;
    extensions?: any;
    headers: Headers;
    status: number;
  }

function queryArgValue(schema: any, arg: any): ValueNode {
  if (arg === null) {
    return {kind: 'NullValue'}
  }

  switch (schema) {
    case 'boolean':
      return {kind: 'BooleanValue', value: arg};
    case 'string':
      return {kind: 'StringValue', value: arg};
    case 'float':
      return {kind: 'FloatValue', value: arg};
    case 'int':
      return {kind: 'IntValue', value: arg};
    default:
      switch (schema.type) {
        case 'boolean':
          return {kind: 'BooleanValue', value: arg};
        case 'string':
          return {kind: 'StringValue', value: arg};
        case 'float':
          return {kind: 'FloatValue', value: arg};
        case 'int':
          return {kind: 'IntValue', value: arg};
      }
  }

  if (arg instanceof Variable) {
    return (
      {
        kind: 'Variable',
        name: {
          kind: 'Name',
          value: arg.name
        }
      }
    )
  }

  throw new Error("Unknown arg value");
}

//   | ListValueNode
//   | ObjectValueNode;

function queryArgs(schema: any, args: any): [ArgumentNode[] | undefined, VariableNode[]] {
  const variables: VariableNode[] = [];
  if (args) {
    const argNodes = Object.entries<any>(args).map(
      ([name, arg]): ArgumentNode => 
        (
          {
            kind: 'Argument',
            name: {kind: 'Name', value: name},
            value: queryArgValue(schema.argFields[name], arg)
          }
        )
    );

    return [argNodes, variables];
  } else {
    return [undefined, []]
  }
}

function queryFields(schema: any, queryFields: any): FieldNode[] {
  return Object.entries<any>(queryFields).map(
    ([name, field]): FieldNode => {
      const [args, variables] = queryArgs(schema[name], field.args)
      const fieldNode: FieldNode = (
        {
          kind: 'Field',
          name: {kind: 'Name', value: field.alias || name},
          arguments: args
        }
      );

      return fieldNode;
    }
  );
}

export function query<Schema extends ValidateSchema<Schema, any>, Query extends GenerateQuery<Schema>>(
  schema: Schema, 
  query: Query, 
  queryName?: string
): DocumentNode {
  const queryNameNode: NameNode | undefined = queryName ? {kind: 'Name', value: queryName} : undefined;
  return ({
    kind: 'Document',
    definitions: [
      {
        kind: 'OperationDefinition',
        operation: 'query',
        name: queryNameNode,
        selectionSet: {
          kind: 'SelectionSet',
          selections: queryFields(schema['queries'], query)
        }
      }
    ]
  });
}

// export type GenerateClientType<
//   Schema extends ValidateSchema<Schema, any>, 
//   Query extends GenerateQuery<Schema>, 
//   Result extends GenerateResult<Schema, Query>
// > = {
//   request: (query: Query) => Promise<Result>,
//   rawRequest: (query: Query) => Promise<RawReturn<Result>>
// }

// export function client<
//   Query extends GenerateQuery<Schema>, 
//   Result extends GenerateResult<Schema, Query>, 
//   Schema extends ValidateSchema<Schema, any> = any
// >(
//   schema: Schema,
//   url: string,
//   options?: RequestInit
// ): GenerateClientType<Schema, Query, Result> {
//   const client = new GraphQLClient(url, options);
//   async function request(queryObject: Query, name?: string): Promise<Result>{
//     const query = createGraphQLQuery(schema, queryObject, name);
//     return await client.request(query);
//   }

//   async function rawRequest(queryObject: Query, name?: string): Promise<RawReturn<Result>>{
//     const query = createGraphQLQuery(schema, queryObject, name)
//     return await client.rawRequest(query);
//   }

//   return ({
//     request,
//     rawRequest
//   });
// }