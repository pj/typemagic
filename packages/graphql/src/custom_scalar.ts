import { GraphQLScalarType } from "graphql";
import { Exact } from "./types";

export class CustomScalar<Type> {
  constructor(
    public scalar: GraphQLScalarType,
  ) {}
}

export function customScalar<Type = never>(scalar: GraphQLScalarType) {
  return new CustomScalar<Type>(scalar);
}

export type HandleCustomScalar<ScalarType, ReturnType> =
  [Exact<ScalarType, ReturnType>] extends [true]
    ? {type: CustomScalar<ReturnType>}
    : "Scalar must have a type specified equal to the return type"
