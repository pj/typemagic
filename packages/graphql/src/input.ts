import { CustomScalar } from "./custom_scalar"
import {
  CreateSchemaOptions,
  Exact,
  GetSchemaScalar,
  GetUnderlyingType,
  IsNonNullNonArrayTypeScalar,
  IsTypeScalar
} from "./types"

export type HandleNonNullNonArrayTypeScalar<Scalar, ArgsRuntimeTypes, Resolver> =
  [IsNonNullNonArrayTypeScalar<Scalar>] extends [true]
    ? GetSchemaScalar<Scalar> | Resolver
    : Resolver

export type ValidateInputRuntimeType<FunctionArg, ArgsRuntimeType> =
  {
    description?: string,
    deprecationReason?: string,
    defaultValue?: string,
  }
  & (
    ArgsRuntimeType extends {enum: infer Enum, name: infer Name, description?: infer Description}
      ? {
          type: {
            enum: Enum, 
            name: Name, 
            description?: Description,
          }
        }
      : ArgsRuntimeType extends {type: CustomScalar<infer CustomScalarType>}
        ? Exact<CustomScalarType, FunctionArg> extends true
          ? {type: CustomScalar<CustomScalarType>}
          : "Custom scalar type doesn't match arg type"
        : IsTypeScalar<FunctionArg> extends true
          ? { type: GetSchemaScalar<FunctionArg> }
          : {
              type: {
                fields: ValidateInputRuntimeTypes<GetUnderlyingType<FunctionArg>, ArgsRuntimeType>,
                name: string
              }
            }
    )
  & CreateSchemaOptions<FunctionArg>

export type ValidateInputRuntimeTypes<FunctionArgs, ArgsRuntimeTypes> =
  {
    [Key in keyof FunctionArgs]:
      HandleNonNullNonArrayTypeScalar<
        FunctionArgs[Key], 
        ArgsRuntimeTypes,
        Key extends keyof ArgsRuntimeTypes
          ? ValidateInputRuntimeType<FunctionArgs[Key], ArgsRuntimeTypes[Key]>
          : ValidateInputRuntimeType<FunctionArgs[Key], unknown>
      >
  }

export type ValidateArgs<FunctionArgs, ArgsRuntimeTypes> =
  [unknown] extends [FunctionArgs]
    ? { argsFields?: never }
    : {
        argsFields: ValidateInputRuntimeTypes<FunctionArgs, ArgsRuntimeTypes>
      }
