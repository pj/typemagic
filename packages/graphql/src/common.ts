import { HandleNonNullNonArrayTypeScalar } from "./resolvers"
import {
  CreateSchemaOptions,
  GetSchemaScalar,
  GetUnderlyingType,
  IsTypeScalar
} from "./types"

export type ValidateInputRuntimeType<FunctionArg> =
  {
    description?: string,
    defaultValue?: string,
  }
  & (
      [IsTypeScalar<FunctionArg>] extends [true]
        ? { type: GetSchemaScalar<FunctionArg> }
        : {
          inputFields: ValidateInputRuntimeTypes<GetUnderlyingType<FunctionArg>>,
          inputName: string
        }
    )
  & CreateSchemaOptions<FunctionArg>


export type ValidateInputRuntimeTypes<FunctionArgs> =
  {
    [Key in keyof FunctionArgs]:
      HandleNonNullNonArrayTypeScalar<FunctionArgs[Key], ValidateInputRuntimeType<FunctionArgs[Key]>>
  }

export type ValidateArgs<FunctionArgs> =
  [unknown] extends [FunctionArgs]
    ? { argsFields?: never }
    : {
        argsFields: ValidateInputRuntimeTypes<FunctionArgs>
      }

export function argsFields<FunctionArgs extends ValidateArgs<FunctionArgs>>(args: FunctionArgs) {
  return args;
}