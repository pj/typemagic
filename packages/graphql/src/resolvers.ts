import { ValidateArgs } from "./common";
import {
  CreateSchemaOptions, Exact, GetSchemaScalar,
  GetUnderlyingType, IsNonNullNonArrayTypeScalar, IsTypeScalar, RemovePromise, SchemaTypeToType
} from "./types";

export type ValidateResolver<Resolver, Root, RootFieldType, Context> =
  [Resolver] extends [{
    type?: infer Type,
    // enum?: infer Enum,
    alias?: infer Alias,
    description?: infer Description,
    argsFields?: infer ArgsRuntimeTypes,
    resolve?: infer ResolverFunction
  }]
    ? { description?: Description, alias?: Alias }
      & ScalarOrObjectType<ReturnTypeForRoot<ResolverFunction, RootFieldType>, GetObjectFields<Type>, Context, Type>
      & CreateSchemaOptions<ReturnTypeForRoot<ResolverFunction, RootFieldType>>
      & ValidateResolverFunction<
          ResolverFunction, 
          Root, 
          ReturnTypeForRoot<ResolverFunction, RootFieldType>, 
          GetObjectFields<Type>, 
          Context,
          Type
        >
    : [IsNonNullNonArrayTypeScalar<RootFieldType>] extends [true]
      ? GetSchemaScalar<RootFieldType>
      : "Can't infer resolver type"

export type ValidateAdditionalResolver<Resolver, Root, Context> =
  [Resolver] extends [{
    type?: infer Type
    // enum?: infer Enum,
    alias?: infer Alias,
    description?: infer Description,
    deprecationReason?: infer DeprecationReason,
    argsFields?: infer ArgsRuntimeTypes,
    resolve?: infer ResolverFunction
  }]
    ? [ResolverFunction] extends [(...args: infer X) => infer RT]
      ? { description?: Description, deprecationReason?: DeprecationReason, alias?: Alias }
        & ScalarOrObjectType<RT, GetObjectFields<Type>, Context, Type>
        & CreateSchemaOptions<RT>
        & ValidateResolverFunction<
            ResolverFunction, 
            Root, 
            RT, 
            GetObjectFields<Type>, 
            Context,
            Type
          >
      : "Resolve function is required on additional fields"
    : "Resolve function is required on additional fields"

// export function resolver<Root, Context, Resolver extends ValidateResolver<Resolver, Root, Context>>(
//   root: Root | Constructor<Root>,
//   context: Context | Constructor<Context>,
//   resolver: Resolver
// ) {
//   return resolver;
// }

export type ValidateResolverFunction<ResolverFunction, Root, RootFieldType, ObjectFields, Context, Type> =
  // ScalarOrObjectType<RootFieldType, ObjectFields, Context, Type>
  // & CreateSchemaOptions<RootFieldType>
  // & 
  (
      [unknown] extends [ResolverFunction]
        ? { resolve?: never, argsFields?: never}
        : [ResolverFunction] extends [(rootOrArgs: infer RootOrArgs, rootOrContext: infer RootOrContext, context: infer X) => infer ReturnType]
          ? [unknown] extends [RootOrArgs]
            ? { 
                argsFields?: never,
                resolve: ((root: Root, context: Context) => RootFieldType) 
                        | ((root: Root, context: Context) => Promise<RootFieldType>)
              }
            : [Exact<RootOrArgs, Root>] extends [true]
              ? { 
                  argsFields?: never,
                  resolve: ((root: Root, context: Context) => RootFieldType) 
                          | ((root: Root, context: Context) => Promise<RootFieldType>)
                }
              : ValidateArgs<RootOrArgs>
                & { 
                    resolve: ((args: RootOrArgs, root: Root, context: Context) => RootFieldType) 
                            | ((args: RootOrArgs, root: Root, context: Context) => Promise<RootFieldType>)
                  }
          : {resolve: "Invalid resolver"} 
      )

export type UnionTypeNames<UnionType> = 
  UnionType extends {objectName: infer Name} ? Name : never


export type ScalarOrObjectType<RootFieldType, ObjectFields, Context, Type> =
  [Type] extends [{enum: infer Enum, name: infer Name, description?: infer Description}]
    ? {type: {enum: Enum, name: Name, description?: Description}}
    : [IsTypeScalar<RootFieldType>] extends [true]
      ? {
          type: GetSchemaScalar<RootFieldType>
        }
        : Type extends {unionName: infer UnionName, unionTypes: infer UnionTypes}
          ? [UnionTypes] extends [Readonly<unknown[]>]
            ? UnionItemToReturnType<UnionTypes[number]> extends infer Unionized
              ? Exact<Unionized, RootFieldType> extends true
                ? {type: {unionName: UnionName, unionTypes: UnionTypes, resolveType?: (value: RootFieldType) => UnionTypeNames<UnionTypes[number]>}}
                : {unionName: UnionName, type: "Union type does not match return type"}
              : never
            : {type: {unionName: UnionName, unionTypes: "Union types should be an array"}}
          : {
              type: {
                objectName: string,
                description?: string,
                deprecationReason?: string,
                objectFields: {
                  [Key in keyof ObjectFields]:
                    [Key] extends [keyof GetUnderlyingType<RootFieldType>]
                      ? ValidateResolver<
                          ObjectFields[Key],
                          GetUnderlyingType<RootFieldType>, 
                          GetUnderlyingType<RootFieldType>[Key], 
                          Context
                        >
                        // Additional properties on object fields are possible.
                      : ValidateAdditionalResolver<
                          ObjectFields[Key],
                          GetUnderlyingType<RootFieldType>, 
                          Context
                        >
                }
              }
            }

export type HandleNonNullNonArrayTypeScalar<Scalar, Resolver> =
  [IsNonNullNonArrayTypeScalar<Scalar>] extends [true]
    ? GetSchemaScalar<Scalar> | Resolver
    : Resolver

// export type NonNullEnum<RootFieldType, Enum, EnumType> =
//   [unknown] extends [Enum] 
//     ? EnumType
//     : [IsNonNullCompileTimeScalar<RootFieldType>] extends [true]
//         ? {enum: Enum} | EnumType
//         : EnumType

export type ReturnTypeForRoot<ResolverFunction, RootFieldType> =
  [unknown] extends [RootFieldType]
    ? [ResolverFunction] extends [(...args: infer Args) => infer ReturnType]
      ? RemovePromise<ReturnType>
      : "Unable to determine return type for root query"
    : RootFieldType

export type GetObjectFields<Type> = 
  [Type] extends [{ objectFields: infer ObjectFields }] 
    ? ObjectFields 
    : unknown
    // : [Type] extends [{enum: infer Enum}] 
    //   ? Type
    //   : unknown

export type NormalEnum<RootFieldType, Enum> = 
  { 
    type: {
      enum: Enum 
    }
  } & CreateSchemaOptions<RootFieldType>

export type HandleEnum<RootFieldType, Type, Enum, NotEnum> =
  [unknown] extends [Enum]
    ? [Type] extends [{enum: infer TypeEnum}] 
      ? NormalEnum<RootFieldType, TypeEnum>
      : NotEnum
    : [IsNonNullNonArrayTypeScalar<RootFieldType>] extends [true]
        ? {enum: Enum} | NormalEnum<RootFieldType, Enum>
        : NormalEnum<RootFieldType, Enum>

export type UnionItemToReturnType<Item> =
  Item extends {objectFields: infer Fields}
    ? {
      [Key in keyof Fields]: SchemaTypeToType<Fields[Key]>
    }
    : never

export function resolver<
  Context, 
  Root, 
  RootField, 
  X extends ValidateResolver<X, Root, RootField, Context> = any
>(resolver: X) {
  return resolver;
}