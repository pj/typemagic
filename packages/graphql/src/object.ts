import { ValidateAdditionalResolver, ValidateResolver } from "./resolvers"
import { GetUnderlyingType, ScalarTypes } from "./types"

export type HandleOutputObject<Type, ReturnType, Context> =
  Type extends {
      name: infer Name,
      description?: string,
      fields: infer ObjectFields,
      interfaces?: infer Interfaces
    }
      ? {
          name: Name,
          description?: string,
          fields: {
            [Key in keyof ObjectFields]:
              [Key] extends [keyof GetUnderlyingType<ReturnType>]
                ? ValidateResolver<
                    ObjectFields[Key],
                    GetUnderlyingType<ReturnType>, 
                    GetUnderlyingType<ReturnType>[Key], 
                    Context
                  >
                : ValidateAdditionalResolver<
                    ObjectFields[Key],
                    GetUnderlyingType<ReturnType>, 
                    Context
                  >
          }
        } 
          & (
            unknown extends Interfaces
              ? {}
              : {
                  interfaces: {
                    [Key in keyof Interfaces]: 
                      ReturnTypeExtendsInterface<Interfaces[Key], ReturnType, Context>
                  }
                }
          )
      : "Unable to infer from object type"
      // {
      //     name: string, 
      //     fields: {
      //       [Key in keyof GetUnderlyingType<ReturnType>]: 
      //         ScalarTypes
      //     },
      //     description?: string,
      //     interfaces?: Array<{[Key in keyof GetUnderlyingType<ReturnType>]: ScalarTypes}>
      //   }

export type ReturnTypeExtendsInterface<Interface, ReturnType, Context> =
  Interface extends {name: infer Name, fields: infer Fields}
      ? {
          name: Name,
          description?: string,
          // Fixme: Can we infer all the possible names this could be?
          resolveType?: (value: ReturnType) => string,
          fields: {
            [Key in keyof Fields]:
              Key extends keyof GetUnderlyingType<ReturnType>
                ? ValidateResolver<
                    Fields[Key],
                    GetUnderlyingType<ReturnType>, 
                    GetUnderlyingType<ReturnType>[Key], 
                    Context
                  >
                : "Interface can't have fields that aren't in return type."
          }
        }
      : "Interface is incorrect"