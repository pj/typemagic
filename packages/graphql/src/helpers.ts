import { ValidateInputRuntimeType } from "./input";
import { HandleOutputObject } from "./object";
import { ValidateResolver } from "./resolvers";

export type ValidateFields<Fields, Root, Context> =
  {
    [Key in keyof Fields]:
      Key extends keyof Root 
        ? ValidateResolver<Fields[Key], Root, Root[Key], Context>
        : ValidateResolver<Fields[Key], Root, unknown, Context>
  }

export type Decircularize<I> = {
  [Key in keyof I]: I
}

export function fields<
  Fields extends ValidateFields<Fields, unknown, unknown>
>(fields: Fields) {
  return fields;
}

export function query<
  Fields extends ValidateResolver<Fields, unknown, unknown, unknown>
>(resolver: Fields) {
  return resolver;
}

export function withContext<Context>() {
  return ({
    fields<Fields extends ValidateFields<Fields, unknown, Context>>(f: Fields) {
      return f
    },
    field<Field extends ValidateResolver<Field, unknown, unknown, Context>>(f: Field) {
      return f
    },
    withType<Type>() {
      return {
        object<Obj extends HandleOutputObject<Decircularize<Obj>, Type, Context>>(obj: Obj) {
          return obj
        },
      }
    },
    withRoot<Root>() {
      return {
        fields<Fields extends ValidateFields<Fields, Root, Context>>(f: Fields) { return f},
      }
    },
    withRootKey<Root, RootKey extends keyof Root>() {
      return {
        field<Field extends ValidateResolver<Field, Root, RootKey, Context>>(f: Field) {return f}
      }
    }
  });
}

export function typeForObject<Type>() {
  return {
    object<Obj extends HandleOutputObject<Decircularize<Obj>, Type, unknown>>(obj: Obj) {
      return obj
    },
    // input<Input extends ValidateInputRuntimeType<Type>>(input: Input) {
    //   return input;
    // }
  }
}

export function withRoot<Root>() {
  return ({
    fields: <Fields extends ValidateFields<Fields, Root, unknown>>(f: Fields) => f,
    field: <Field extends ValidateResolver<Field, Root, unknown, unknown>>(f: Field) => f,
    withContext: <Context>() => ({
      fields:<Fields extends ValidateFields<Fields, Root, Context>>(f: Fields) => f,
      field: <Field extends ValidateResolver<Field, Root, unknown, Context>>(f: Field) => f
    }),
  })
}

export function withRootKey<Root, RootKey extends keyof Root>() {
  return ({
    field: <Field extends ValidateResolver<Field, Root, RootKey, unknown>>(f: Field) => f,
  })
}

export function withContextRoot<Context, Root>() {
  return ({
    fields<Fields extends ValidateFields<Fields, Root, Context>>(f: Fields) {
      return f
    },
    additionalField<Field extends ValidateResolver<Field, Root, unknown, Context>>(f: Field) {
      return f
    }
  });
}

export function fieldContextAndKey<Context, Root, RootKey extends keyof Root>() {
  return ({
    field<Field extends ValidateResolver<Field, Root, RootKey, Context>>(f: Field) {
      return f
    }
  });
}