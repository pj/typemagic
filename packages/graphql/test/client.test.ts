import { ScalarTypes } from "../src";
import { query, _ } from "../src/client";

test('clientASTGeneration', async () => {
  expect(
    query(
      {
        queries: {
          testQuery: {
            type: 'string',
            resolve: () => {
              return 'asdf';
            }
          }
        }
      },
      {
        testQuery: _
      }
    )
  ).toMatchSnapshot();
});