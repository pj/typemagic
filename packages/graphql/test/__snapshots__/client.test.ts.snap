// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clientASTGeneration 1`] = `
Object {
  "definitions": Array [
    Object {
      "kind": "OperationDefinition",
      "name": undefined,
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "arguments": undefined,
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "testQuery",
            },
          },
        ],
      },
    },
  ],
  "kind": "Document",
}
`;
